### ISSUE TITLE
could not convert ‘qh_qh’ from ‘qhT’ to ‘bool’

### ISSUE THREAD

**Issue by dpettas** [schansge/libdistmesh#3]
Hello, 
I was trying to compile the library using g++ on Ubuntu 18.04. 
However, at the triangulation.cpp I get this error:

error: could not convert ‘qh_qh’ from ‘qhT’ to ‘bool.’ 

Do you know how to fix it? 

Thank you,

**Comment by krober10nd** [comment]
Yes you haven't included the correct paths to qhull of you haven't built qhull correctly. Please see the makefile and check the paths are correct.

**Comment by stephenmbull** [comment]
Hello,
I'm having the darndest time with the same issue. @dpettas were you able to resolve this? I'm also running Ubuntu 18.04. I'm perhaps missing something obvious.

I've checked my qhull install and paths, and other than a minor include correction in triangulation.cpp (#include <qhull/qhull_a.h> >> #include <libqhull/qhull_a.h) I'm still somewhat stuck. 

My main concern is use of the latest qhull (currently 2020.2.r 2020/07/24). The qhull docs suggest that everything should be backwards compatible, at least back to the 2012.1 era.

Any further insights would be appreciated. I'm going to continue hunting around and possibly try an older version of qhull.

Thanks much!

**Comment by stephenmbull** [comment]
Hello again,

Okay, issue resolved. The newer versions of Qhull made a minor architectural change that is reflected in their change logs and a couple of header files. 

Note: I am using Qhull 2020.2.r 2020/07/24

To resolve this error edit libqhull/user.h (may need to be sudo). Search for qh_QHpointer. The doc notes in here explain its usage (which are also referred to in the change logs). Change the default defined setting of 0 to 1:

#define qh_QHpointer 0 >> #define qh_QHpointer 1 

This will also resolve the second make error for triangulation.cpp, " ‘qh_save_qhull’ was not declared in this scope "
qh_save_qhull is in libqhull/libqhull.h and its definition is dependent upon the above qh_QHpointer being 1.

Thanks,
Stephen

**Comment by yjhp1016** [comment]
Hello Stephen,
My qh_QHpointer has been set as 1, but sill can't compile distmesh, I got same error as @dpettas ... Do you have any idea how to solve it? 

Thanks!

Paul

**Comment by pgebhardt** [comment]
Hi guys,

I haven't used the lib for a long time now, so I am sure the best way to proceed here would be to replace the calls to the hull C interface with calls to the newer C++ Interface. However, I checked on macOS Big Sur and it works by simply commenting out lines 35-37 in src/triangulation.cpp.

Idk if it also helps on other platforms, but go ahead and give it a try.

Best, Patrik

**Comment by pgebhardt** [comment]
Ok,

I gave it another try on Ubuntu 20.04 and installing libeigen3-dev and libqhull-dev using apt from upstream ubuntu sources. After adding the necessary include paths to Makefile.config it worked right away, without tampering with any source files of libdistmesh or libqhull.

Could you tell me more about your build environment so I can try to replicate your errors more easily?

Thanks and best, Patrik

**Comment by yjhp1016** [comment]
Hello Patrik,

Thanks a lot for your quick response, I'm using UBuntu 20.04 as well, I
compile eigen or qhull from source, but I just tried to install
libeigen3-dev and libqhull using apt, run it again, found similar error:
src/triangulation.cpp: In function ‘Eigen::ArrayXXi
distmesh::triangulation::delaunay(Eigen::Ref<const Eigen::Array<double, -1,
-1> >)’:
src/triangulation.cpp:35:9: error: could not convert ‘qh_qh’ from ‘qhT’ to
‘bool’
   35 |     if (qh_qh) {
      |         ^~~~~
      |         |
      |         qhT
src/triangulation.cpp:36:9: error: ‘qh_save_qhull’ was not declared in this
scope; did you mean ‘qh_freeqhull’?
   36 |         qh_save_qhull();
      |         ^~~~~~~~~~~~~
      |         qh_freeqhull
make: *** [Makefile:123:
build/release/x86_64-linux-gnu/objs/src/triangulation.o] Error 1

But it is working by commenting out the lines 35-37. If these lines are not
super important, I will just keep them commented out :-)

Thank you very much!

Paul
